rs/replicated_state/src/canister_state/queues.rs

```diff
@@ -916,20 +898,6 @@ impl CanisterQueues {
             return;
         }
 
-        // Adjust stats for any queue pairs we are going to GC.
-        let mut have_empty_pair = false;
-        for (_canister_id, (input_queue, output_queue)) in self.canister_queues.iter() {
-            if !input_queue.has_used_slots() && !output_queue.has_used_slots() {
-                self.input_queues_stats.size_bytes -= input_queue.calculate_size_bytes();
-                have_empty_pair = true;
-            }
-        }
-        // Bail out early if there is nothing to GC.
-        if !have_empty_pair {
-            return;
-        }
-
-        // Actually drop empty queue pairs.
         self.canister_queues
             .retain(|_canister_id, (input_queue, output_queue)| {
                 input_queue.has_used_slots() || output_queue.has_used_slots()
```

Comment:
Do we also want to check below whether the queue invariants are ok after garbage collection? Possibly even for the default values above?

Reply:
We have `stats_ok()`, which is validated below; `schedules_ok()` (input queue schedule invariants); and `canister_queues_ok()` (non-stale reference at the head of every queue). Which one are you referring to?

I guess one could validate `schedules_ok()`, even though we explicitly allow for empty or missing queues in input schedules. `canister_queues_ok()` is unnecessary, because we don't actually mutate any individual canister queue; or the message pool.

Reply:
I guess what confuses me is that there are multiple different checks and we only call the ones we conclude are relevant based on what we touch. This adds cognitive load because one always needs to know which invariants are relevant and which not. Would it maybe make sense to have a check which combines all three and we just call it at the end of every function that mutates things in any way?

Reply:
In most cases, maybe. But there are situations (e.g. wherever I used the `canister_queue[s]_ok()` functions instead of the method) where the borrow checker will not let you call even immutable methods. I can give it a go.

Reply:
Both of them are, since they're quite expensive checks (`O(n * log(n))` time complexity both).

Technically, it may be safer to use `assert!()` for the hard invariants (so we don't end up persisting a state with a broken hard invariant that we cannot load back afterwards). But some of those checks (e.g. that `input_schedule_canisters` contains exactly the canisters scheduled in the input schedules) are expensive to do on every mutation. Plus, calling the `debug_asserts` on every mutation should make it virtually impossible for a bug to slip by unnoticed into production code.

Reply:
Ok. Thanks for the clarification and for the change.